rama-net/src/address/domain_address.rs

```diff
@@ -0,0 +1,198 @@
+
+    fn try_from(s: &str) -> Result<Self, Self::Error> {
+        let (domain, port) = parse_utils::split_port_from_str(s)?;
+        let domain = Domain::from_str(domain)?;
+        Ok(Self::new(domain, port))
+    }
+}
+
+impl TryFrom<String> for DomainAddress {
+    type Error = OpaqueError;
+
+    fn try_from(s: String) -> Result<Self, Self::Error> {
+        let (domain, port) = parse_utils::split_port_from_str(&s)?;
+        let domain = Domain::from_str(domain)?;
+        Ok(Self::new(domain, port))
+    }
+}
+impl TryFrom<Vec<u8>> for DomainAddress {
+    type Error = OpaqueError;
+
+    fn try_from(bytes: Vec<u8>) -> Result<Self, Self::Error> {
+        let s = String::from_utf8(bytes).context("parse domain_address from bytes")?;
+        s.try_into()
+    }
+}
+
+impl TryFrom<&[u8]> for DomainAddress {
+    type Error = OpaqueError;
+
+    fn try_from(bytes: &[u8]) -> Result<Self, Self::Error> {
+        let s = std::str::from_utf8(bytes).context("parse domain_address from bytes")?;
+        s.try_into()
+    }
+}
+
+impl serde::Serialize for DomainAddress {
+    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
+    where
+        S: serde::Serializer,
+    {
+        let address = self.to_string();
+        address.serialize(serializer)
+    }
+}
+
+impl<'de> serde::Deserialize<'de> for DomainAddress {
+    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
+    where
+        D: serde::Deserializer<'de>,
+    {
+        let s = <&'de str>::deserialize(deserializer)?;
+        s.parse().map_err(serde::de::Error::custom)
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+
+    fn assert_eq(s: &str, domain_address: DomainAddress, domain: &str, port: u16) {
+        assert_eq!(domain_address.domain().as_str(), domain, "parsing: {}", s);
+        assert_eq!(domain_address.port(), port, "parsing: {}", s);
+    }
+
+    #[test]
+    fn test_valid_domain_address() {
+        for (s, (expected_domain, expected_port)) in [
+            ("example.com:80", ("example.com", 80)),
+            ("subdomain.example.com:443", ("subdomain.example.com", 443)),
+        ] {
+            let msg = format!("parsing '{}'", s);
+
+            assert_eq(s, s.parse().expect(&msg), expected_domain, expected_port);
+            assert_eq(s, s.try_into().expect(&msg), expected_domain, expected_port);
+            assert_eq(
+                s,
+                s.to_owned().try_into().expect(&msg),
+                expected_domain,
+                expected_port,
+            );
+            assert_eq(
+                s,
+                s.as_bytes().try_into().expect(&msg),
+                expected_domain,
+                expected_port,
+            );
+            assert_eq(
+                s,
+                s.as_bytes().to_vec().try_into().expect(&msg),
+                expected_domain,
+                expected_port,
+            );
+        }
+    }
+
+    #[test]
+    fn test_parse_invalid() {
+        for s in [
```

Comment:
Can you also add some tests to ensure that valid addresses that use an ip address are still invalid for here. E.g. 127.0.0.1:8080, as well as an ipv6 version

Reply:
Hey! I see `Domain.rs` parses 127.0.0.1 as a valid domain. Will fix this and update code.

Reply:
127.0.0.1 is a valid domain though, odd as it may seem. Might have to just accept that reality.

Reply:
Do I update Domain.rs to not accept 127.0.0.1 as a valid domain? I'm confused here.

Reply:
Nope 127.0.0.1 is valid sadly, I forgot that when creating this ticket. Ipv6 addresses are however not valid.

The good news is that domain 127.0.0.1 will still fail at runtime as that wont dns resolve.

Reply:
You still can add the it fails ti parse [ipv6]:port